src/coins.h

```diff
@@ -130,6 +157,56 @@ struct CCoinsCacheEntry
     CCoinsCacheEntry() : flags(0) {}
     explicit CCoinsCacheEntry(Coin&& coin_) : coin(std::move(coin_)), flags(0) {}
     CCoinsCacheEntry(Coin&& coin_, unsigned char flag) : coin(std::move(coin_)), flags(flag) {}
+    ~CCoinsCacheEntry()
+    {
+        // We must always clear the flags when destroying this to remove it from
+        // the flagged linked list
+        ClearFlags();
+    }
+
+    //! Adding a flag also requires a self reference to the pair that contains
+    //! this entry in the CCoinsCache map and a reference to the head of the
+    //! flagged pair linked list.
+    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& head)
+    {
+        if (flags == 0) return;
+        if (m_flags != 0) {
+            m_flags |= flags;
+            return;
+        }
+        m_prev = &head;
+        m_next = head.second.m_next;
+        head.second.m_next = &self;
+        if (m_next != nullptr) {
```

Comment:
we could move the `m_next` setter closer to the check, something like:
```C++
    inline void AddFlags(uint8_t flags, CoinsCachePair& self, CoinsCachePair& head)
    {
        if (flags && !m_flags) {
            m_prev = &head;
            m_next = head.second.m_next;
            if (m_next) m_next->second.m_prev = &self;
            head.second.m_next = &self;
        }
        m_flags |= flags;
    }
```

Reply:
Done.